homeassistant/components/tuya/__init__.py

```diff
@@ -50,34 +73,90 @@
 
 TUYA_TRACKER = "tuya_tracker"
 
+TUYA_DEVICE_CONF_SCHEMA = {
+    vol.Optional(TUYA_DEVICES_CONF): vol.All(
+        cv.ensure_list,
+        [
+            vol.Schema(
+                {
+                    vol.Required(CONF_DEVICE_NAME): cv.string,
+                    vol.Optional(CONF_UNIT_OF_MEASUREMENT): cv.temperature_unit,
+                    vol.Optional(CONF_TEMP_DIVIDER, default=0): cv.positive_int,
+                    vol.Optional(CONF_CURR_TEMP_DIVIDER, default=0): cv.positive_int,
+                    vol.Optional(CONF_EXT_TEMP_SENSOR): cv.string,
+                    vol.Optional(CONF_SUPPORT_COLOR): cv.boolean,
+                    vol.Optional(
+                        CONF_BRIGHTNESS_RANGE_MODE, default=0
+                    ): cv.positive_int,
+                    vol.Optional(CONF_MIN_KELVIN): cv.positive_int,
+                    vol.Optional(CONF_MAX_KELVIN): cv.positive_int,
+                    vol.Optional(CONF_MAX_TUYA_TEMP): cv.positive_int,
+                }
+            )
+        ],
+    )
+}
+
 CONFIG_SCHEMA = vol.Schema(
     vol.All(
-        cv.deprecated(DOMAIN),
         {
             DOMAIN: vol.Schema(
                 {
-                    vol.Required(CONF_PASSWORD): cv.string,
-                    vol.Required(CONF_USERNAME): cv.string,
-                    vol.Required(CONF_COUNTRYCODE): cv.string,
+                    vol.Optional(CONF_USERNAME): cv.string,
+                    vol.Optional(CONF_COUNTRYCODE): cv.string,
+                    vol.Optional(CONF_PASSWORD): cv.string,
                     vol.Optional(CONF_PLATFORM, default="tuya"): cv.string,
                 }
-            )
+            ).extend(TUYA_DEVICE_CONF_SCHEMA)
```

Comment:
Unfortunately, additions/changes to the yaml schema are no longer allowed in integrations for devices or services. See [ADR0010](https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md#decision).

It is not clear to me how one could offer per-device configuration options like these via config flow, given that you also have settings shared between devices (without the latter one could just do one config entry per device). 

Indeed, supporting device level configuration is something that "should be added", per [@balloob in an architecture discussion](https://github.com/home-assistant/architecture/issues/381#issuecomment-625530066).

Given that these options are currently needed to make certain devices work properly, this does need to be exposed. While handling these via entity customization would be possible, that is not permitted to be used for potentially required things like these. Perhaps an exception ADR0010 should be made here, pending such a per-device configuration system being implemented?

I do wonder how the official Tuya app handles things, given that the rest API fails to expose the needed information in all cases. Does the app use some undocumented APIs? Does it have some database of devices built in? Is it simply broken for some devices? (Seems unlikely.) If we could somehow avoid the need to enable users to manually configure these to have the devices work, that would be ideal.

Reply:
The API that HA via TuyaHA library is using, is a special API that was developed by Tuya some years ago. Recently Tuya decide to limit the access frequency to the call, and this is a creating a lot of issue and error due to fact that the API method is called with very high frequency, especially for user that have a great number of registered device.
Tuya provide an SDK that allow vendor that use tuya Cloud to create dedicated app, but this is not applicable to HA. Tuya and SmartLive app use the same logic, anyway I I have no idea on how this could be applied to HA

Dedicate HA Tuya API also have some limitation that cause some device not working properly because the Cloud do not provide all necessary information, and the configuration.yaml information is used to configured this missing info using some additional method implemented in TuyaHA library with last PR.
In any case I understand that usage of configuration.yaml for new integration should be avoided, but, as you also say, at this moment is not possible to provide configuration for single device using UI.

As alternative approach, I can also remove all device configuration from this PR and implement only the fix that will avoid connection error, the result is that already know limitation for the moment will not be fixed but al least status refresh and control for working device will be fixed.
Please let me know how I should proceed.

Reply:
Yeah, I get all that. I had not noticed this was a special API though. Looking a bit deeper into things they literally have at least 3 different http APIs, which is rather crazy. 

The issues caused by the HA API now having strict rate limits, which can break the integration for people with multiple devices is certainly a high priority thing to merge.

In addition the special HA API fails to return information that is needed, so per device configuration is required to ensure devices can work. Hence the need for per device configuration. This is also an important thing to address.

Personally I do feel like this is something worth making a one-time exception to the ADR for, but that is not my decision to make. Hopefully somebody from the core dev team will comment on this.

Reply:
Rather then implementing additional schema, can this data be collected in a config flow similar to how https://github.com/rospogrigio/localtuya-homeassistant is doing it?

Reply:
I think that local_tuya use different entry_id for every device, while Tuya integration use a single entry_id for multiple device and new registered device in Tuya app are auto discovered. Furthermore implementing multiple entry_id require a refresh call for every device and this would invalidate the optimization work done, apart from this will completely change the integration behavior.
I could also try to implement an initial wizard for all discovered device during integration configuration, but than will be impossible to configure new discovered device without removing and re-adding integration.

Reply:
Technically, one can handle this during editing too. But this is not going to be terribly ergonomic.

To lay out such an approach to be as general as possible (possibly more general than would be needed here):
1. First form(s) is/are normal shared level options.
2. Next form is a menu form that asks what additional action to take: "Add Device Configuration", "Remove Device Configuration", "Edit Device Configuration", "Finished Configuring".
     1. If you pick add, it takes you to a form with device level settings. And when finished, it returns you to the menu form.
     2. If you pick "edit" it shows you a list of devices that currently have configuration. You pick one, it brings you to a form to edit it. Once done, you return to the menu.
     3. If you pick "Remove", you see a list of devices that currently have configuration. You can pick one or more. When you click submit it takes you back to the menu.
     4. If you pick "Finished Configuring", the process ends.

Obviously in between each step you are applying the relevant changes to the in-progress version of the config entry.

One could probably do slightly better if discovered devices are available. In that case, one would know what all the discovered devices are, and could simply list them, along with options for "edit" or "reset". Edit would create a default configuration for said device it it did not have one already, and then show an edit dialog. Reset would simply delete the device level configuration if one exists.

Where things get really messy is if a device configuration needs some form of list of complex values. One would probably need to repeat a similar system menu system at the device level, nested inside the other menu.

Reply:
I will try do do something as suggested. I will report progress...

Reply:
Working on this, I think that adding a combo in ConfigOption flow I can manage this and remove the additional schema. Need some days to complete this change...

Reply:
Finish implementing, this what I did:

1) when you have a configurable device (light or climate) you can select in from a drop-down list (default is save configuration)

![image](https://user-images.githubusercontent.com/60491700/96335023-8562e100-1075-11eb-9de8-388dc6ab33bc.png)

3) If you select a device, open the page for the corresponding type:

Light:

![image](https://user-images.githubusercontent.com/60491700/96335049-bf33e780-1075-11eb-9896-f5d136b32900.png)

Climate:

![image](https://user-images.githubusercontent.com/60491700/96335060-de327980-1075-11eb-8a76-02dcd5918656.png)

3) When you press "save", it come back to the main page, so that you can choose to configure another device or finally save configuration leaving default value "Save Configuration" in drop-down.

What do you think?

Reply:
I don't have a horse in this race, but I'd rename `Save configuration` to something that tells me it's going to apply across devices and not per device.

Reply:
Suggestion?

Reply:
what about `All devices` ? it aligns with the message `Select device ...`

Reply:
But `Save Configuration` means that you are going to "Save the configuration" when you press "Apply". `All device` will not clarify this. May be I do not understand what you have in mind?

Reply:
So you have an option that says: `Select the device to configure` and in the dropdown, you have two options: `Save configuration` and `Lampux-RGBCWA60 (light)`. I was thinking that is better to rename the dropdown option to `All devices`, so when you click done, you know it applies to _all_. 

Not sure if that helps my suggestion more clear 😅